testinfra/utils/__init__.py

```diff
@@ -14,11 +14,42 @@
 from __future__ import unicode_literals
 
 import re
+import six
+import string
 
+if six.PY3:
+    # pylint: disable=no-member
+    _find_unsafe = re.compile(r'[^\w@%+=:,./-]', re.ASCII)
+    # pylint: enable=no-member
+else:
+    _safechars = frozenset(string.ascii_letters + string.digits + '@%_-+=:,./')
 first_cap_re = re.compile('(.)([A-Z][a-z]+)')
 all_cap_re = re.compile('([a-z0-9])([A-Z])')
 
 
 def un_camel_case(name):
     s1 = first_cap_re.sub(r'\1_\2', name)
     return all_cap_re.sub(r'\1_\2', s1).lower()
+
+
+# (gtmanfred) This is copied from pipes.quote, but changed to use double quotes
+# instead of single quotes.  This is used by the winrm backend.
+def quote(s):
+    """Return a shell-escaped version of the string *s*."""
+    if not s:
+        return "''"
+    if six.PY3:
+        if _find_unsafe.search(s) is None:
+            return s
+    else:
+        for c in s:
+            if c not in _safechars:
+                break
+        else:
+            if not s:
+                return "''"
+            return s
+
+    # use single quotes, and put single quotes into double quotes
+    # the string $'b is then quoted as '$'"'"'b'
+    return '"' + s.replace('"', '"\'"\'"') + '"'
```

Comment:
Can you put this code in winrm.py since it's used only for this backend ?

Also it will be great to add relevant tests for this.

Reply:
added.